homeassistant/components/swiss_public_transport/sensor.py

```diff
@@ -107,60 +95,33 @@ async def async_setup_platform(
         )
 
 
-class SwissPublicTransportSensor(SensorEntity):
+class SwissPublicTransportSensor(
+    CoordinatorEntity[SwissPublicTransportDataUpdateCoordinator], SensorEntity
+):
     """Implementation of an Swiss public transport sensor."""
 
     _attr_attribution = "Data provided by transport.opendata.ch"
     _attr_icon = "mdi:bus"
 
-    def __init__(self, opendata: OpendataTransport, name: str) -> None:
+    def __init__(
+        self, coordinator: SwissPublicTransportDataUpdateCoordinator, name: str
+    ) -> None:
         """Initialize the sensor."""
-        self._opendata = opendata
+        super().__init__(coordinator)
+        self._coordinator = coordinator
         self._attr_name = name
         self._remaining_time: timedelta | None = None
 
     @property
     def native_value(self) -> str:
         """Return the state of the sensor."""
-        return self._opendata.connections[0]["departure"]
+        return self.coordinator.data["departure"]
 
     @property
     def extra_state_attributes(self) -> Mapping[str, Any]:
         """Return the state attributes."""
-        departure_time = dt_util.parse_datetime(
-            self._opendata.connections[0]["departure"]
-        )
-        if departure_time:
-            remaining_time = departure_time - dt_util.as_local(dt_util.utcnow())
-        else:
-            remaining_time = None
-        self._remaining_time = remaining_time
-
         return {
-            ATTR_TRAIN_NUMBER: self._opendata.connections[0]["number"],
-            ATTR_PLATFORM: self._opendata.connections[0]["platform"],
-            ATTR_TRANSFERS: self._opendata.connections[0]["transfers"],
-            ATTR_DURATION: self._opendata.connections[0]["duration"],
-            ATTR_DEPARTURE_TIME1: self._opendata.connections[1]["departure"],
-            ATTR_DEPARTURE_TIME2: self._opendata.connections[2]["departure"],
-            ATTR_START: self._opendata.from_name,
-            ATTR_TARGET: self._opendata.to_name,
-            ATTR_REMAINING_TIME: f"{self._remaining_time}",
-            ATTR_DELAY: self._opendata.connections[0]["delay"],
+            key: value
+            for key, value in self.coordinator.data.items()
+            if key not in {"departure"}
```

Comment:
Instead of calculating this each time this property is accessed, we can override the `_handle_coordinator_update` from the parent class and pre-calculate it and store it in `_attr_extra_state_attributes`.

Alternative, separate them in the DataUpdateCoordinator, so the calculation only happens once every update.

Reply:
Ok, will hava a look at that function👍

Reply:
To be honest, I think the latter is the best option. As you are already handling data in the coordinator.

You could as well use a tuple of dict there, that sets separate values for state & extra attributes in the coordinator data.

Reply:
I think I would like to get rid of the attributes in the long run and therefore not touch the data coordinator. Instead contain the attribute handling in the sensor and in a follow up PR more sensors (after it has become also a device).

Reply:
So I will try the handle function

Reply:
@frenck I tried now both methods, and I think the `_handle_coordinator_update` is cleaner, especially considering the extra attributes should get deprecated in the future. The data coordinator stays lean and clean this way :)